fadecandy_driver/src/fadecandy_driver.cpp

```diff
@@ -0,0 +1,301 @@
+    {
+      fadecandy_device = device;
+      fadecandy_device_descriptor = desc;
+      //      printf("Vendor:Device = %04x:%04x\n", desc.idVendor,
+      //      desc.idProduct);
+    }
+  }
+}
+
+std::vector<std::vector<unsigned char>>
+FadecandyDriver::makeVideoUsbPackets(std::vector<std::vector<colors>> led_array_colors)
+{
+  int led_index = 0;
+  std::vector<colors> all_led_colors(LEDS_PER_STRIP * NUM_STRIPS, { 0, 0, 0 });
+  for (size_t i = 0; i < led_array_colors.size(); i++)
+  {
+    for (size_t j = 0; j < led_array_colors[i].size(); j++)
+    {
+      led_index = (i * LEDS_PER_STRIP) + j;
+      all_led_colors[led_index] = led_array_colors[i][j];
+    }
+  }
+
+  assert(all_led_colors.size() == LEDS_PER_STRIP * NUM_STRIPS);
+  std::vector<std::vector<unsigned char>> packets;
+  std::vector<unsigned char> packet;
+  std::vector<colors> packet_leds;
+  std::vector<colors> remaining_leds = all_led_colors;
+  std::vector<int> color_bytes;
+  int control;
+
+  while (remaining_leds.size() > 0)
+  {
+    color_bytes.clear();
+    packet.clear();
+
+    if (remaining_leds.size() < LEDS_PER_PACKET)
+    {
+      packet_leds.assign(remaining_leds.begin(), remaining_leds.end());
+      remaining_leds.erase(remaining_leds.begin(), remaining_leds.end());
+    }
+    else
+    {
+      packet_leds.assign(remaining_leds.begin(), remaining_leds.begin() + LEDS_PER_PACKET);
+      remaining_leds.assign(remaining_leds.begin() + LEDS_PER_PACKET, remaining_leds.end());
+    }
+
+    control = packets.size() | PACKET_TYPE_VIDEO;
+    if (remaining_leds.size() == 0)
+      control |= FINAL_PACKET_BIT;
+
+    for (size_t i = 0; i < packet_leds.size(); i++)
+    {
+      color_bytes.push_back(packet_leds[i].r);
+      color_bytes.push_back(packet_leds[i].g);
+      color_bytes.push_back(packet_leds[i].b);
+    }
+
+    if (63 - color_bytes.size() > 0)
+    {
+      int j = 63 - color_bytes.size();
+      for (int i = 0; i < j; i++)
+      {
+        color_bytes.push_back(0);
+      }
+    }
+
+    for (size_t i = 0; i < color_bytes.size(); i++)
+    {
+      int byts_per_int = 1;
+      unsigned char buffer[byts_per_int];
+      std::vector<unsigned char> bufferv;
+      bufferv = intToCharArray(buffer, color_bytes[i], byts_per_int);
+      packet.insert(packet.end(), bufferv.begin(), bufferv.end());
+    }
+
+    // add control byte
+    packet.insert(packet.begin(), static_cast<unsigned char>(control));
+
+    assert(packet.size() == USB_PACKET_SIZE);
+    packets.push_back(packet);
+  }
+  return packets;
+}
+
+std::vector<std::vector<unsigned char>> FadecandyDriver::makeLookupTablePackets(std::vector<int> red_lookup_values,
+                                                                                std::vector<int> green_lookup_values,
+                                                                                std::vector<int> blue_lookup_values)
+{
+  assert(red_lookup_values.size() == LOOKUP_VALUES_PER_CHANNEL);
```

Comment:
Please throw std exceptions instead of these asserts if you would like to do runtime checking